core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -59,24 +61,101 @@ object RequestChannel extends Logging {
     val sanitizedUser: String = Sanitizer.sanitize(principal.getName)
   }
 
-  class Metrics(enabledApis: Iterable[ApiKeys]) {
-    def this(scope: ListenerType) = {
-      this(ApiKeys.apisForListener(scope).asScala)
+  class Metrics(enabledApis: Iterable[ApiKeys], requestMetricsSizeBucketsConfig: String) {
+    def this(scope: ListenerType, requestMetricsSizeBucketsConfig: String) = {
+      this(ApiKeys.apisForListener(scope).asScala, requestMetricsSizeBucketsConfig)
     }
 
     private val metricsMap = mutable.Map[String, RequestMetrics]()
 
+    // map[acks, map[requestSize, metricName]]
+    // this is used to create metrics for produce requests of different size and acks
+    val produceRequestAcksSizeMetricNameMap = getProduceRequestAcksSizeMetricNameMap()
+    // map[responseSize, metricName]
+    // this is used to create metrics for fetch requests of different response size
+    val consumerFetchRequestSizeMetricNameMap = getConsumerFetchRequestAcksSizeMetricNameMap
+
     (enabledApis.map(_.name) ++
       Seq(RequestMetrics.MetadataAllTopics) ++
-      Seq(RequestMetrics.consumerFetchMetricName, RequestMetrics.followFetchMetricName)).foreach { name =>
-      metricsMap.put(name, new RequestMetrics(name))
+      Seq(RequestMetrics.consumerFetchMetricName, RequestMetrics.followFetchMetricName) ++
+      getConsumerFetchRequestSizeMetricNames ++
+      getProduceRequestAcksSizeMetricNames)
+      .foreach { name => metricsMap.put(name, new RequestMetrics(name))
     }
 
     def apply(metricName: String): RequestMetrics = metricsMap(metricName)
 
     def close(): Unit = {
        metricsMap.values.foreach(_.removeMetrics())
     }
+
+    // generate map[request/responseSize, metricName] for requests of different size
+    def getRequestSizeMetricNameMap(requestType: String):Map[Int, String] = {
+      // convert requestMetricsSizeBucketsConfig from string to int array. e.g., "0,1,10,50,100" => [0,1,10,50,100]
+      var bucketsString = requestMetricsSizeBucketsConfig.replaceAll("\\s", "").split(',')
+      if(bucketsString.length < 2) {
```

Comment:
The sanity check on the config should be moved upfront during broker startup. Below is an example
https://github.com/linkedin/kafka/blob/5472f4897ded846afe62481e988373bf90d4c0e2/core/src/main/scala/kafka/server/KafkaConfig.scala#L1926

Reply:
added validator in .define and removed the check here